### ISSUE TITLE
onOSSubscriptionChanged only gets called when app is closed and reopened after giving granted permissions

### ISSUE THREAD

**Issue by LamourBt** [OneSignal/OneSignal-iOS-SDK#411]
**Description:**
Hey @Nightsd01 
Immediately after the user `allows` push notification, it seems that the device was registered to APNS, however OneSignal Delegates wouldn't be called with the  new data such as the state and player ID  

I encountered the same issue in both `2.8.5` then I update to `2.8.7` (using cocoapods)

Immediately after the user `allows` push notification I get the following within the log 
```
2018-09-08 10:53:05.933515-0400 Pointer[26193:8824462] VERBOSE: oneSignalDidRegisterForRemoteNotifications:deviceToken:
2018-09-08 10:53:05.933697-0400 Pointer[26193:8824462] INFO: Device Registered with Apple: 573cc71dfe6611d977516a2d8093fae3dccb530cc0d94075574cf590dbd7c17d
``` 
**And still `onOSSubscriptionChanged` is still not called.** 

The crazy part is for it `onOSSubscriptionChanged` to be called the app needs to be closed and reopen. here is the some of the log 

```
2018-09-08 10:54:38.363542-0400 Pointer[26209:8826709] INFO: Device Registered with Apple: 573cc71dfe6611d977516a2d8093fae3dccb530cc0d94075574cf590dbd7c17d
SubscriptionStateChange: 
Optional(<OSSubscriptionStateChanges:
from: <OSSubscriptionState: userId: (null), pushToken: (null), userSubscriptionSetting: 1, subscribed: 0>,
to:   <OSSubscriptionState: userId: (null), pushToken: 573cc71dfe6611d977516a2d8093fae3dccb530cc0d94075574cf590dbd7c17d, userSubscriptionSetting: 1, subscribed: 0>
>)
```
Some of the verbose log 
```
2018-09-08 10:59:25.339760-0400 Pointer[26218:8830249] VERBOSE: registerForPushNotifications Called:waitingForApnsResponse: 0
2018-09-08 10:59:25.339934-0400 Pointer[26218:8830249] VERBOSE: Firing registerForRemoteNotifications
RootViewControlller.viewDidLoad.tableview, view did load delegate methods
2018-09-08 10:59:25.554348-0400 Pointer[26218:8830249] VERBOSE: ONESIGNAL setOneSignalDelegate CALLED: (null)
2018-09-08 10:59:25.554412-0400 Pointer[26218:8830249] VERBOSE: ONESIGNAL setOneSignalDelegate CALLED: <Pointer.AppDelegate: 0x1d4459170>
2018-09-08 10:59:25.554669-0400 Pointer[26218:8830249] VERBOSE: updateNotificationTypes called: 15
2018-09-08 10:59:25.554705-0400 Pointer[26218:8830249] VERBOSE: startedRegister: 1
2018-09-08 10:59:25.554746-0400 Pointer[26218:8830249] VERBOSE: getNotificationTypes:mSubscriptionStatus: -1
2018-09-08 10:59:25.556506-0400 Pointer[26218:8830249] VERBOSE: getNotificationTypes:mSubscriptionStatus: -1
2018-09-08 10:59:25.562607-0400 Pointer[26218:8830249] VERBOSE: oneSignalDidRegisterForRemoteNotifications:deviceToken:
2018-09-08 10:59:25.562689-0400 Pointer[26218:8830249] INFO: Device Registered with Apple: 573cc71dfe6611d977516a2d8093fae3dccb530cc0d94075574cf590dbd7c17d
2018-09-08 10:59:25.562814-0400 Pointer[26218:8830249] VERBOSE: updateDeviceToken:onSuccess:onFailure:
2018-09-08 10:59:25.562904-0400 Pointer[26218:8830249] VERBOSE: getNotificationTypes:mSubscriptionStatus: -1
2018-09-08 10:59:25.564827-0400 Pointer[26218:8830249] INFO: Device Registered with OneSignal: 7abcaf0f-f7dc-4945-9f9f-a9370894dc4c
2018-09-08 10:59:25.567937-0400 Pointer[26218:8830249] VERBOSE: oneSignalApplicationDidBecomeActive
2018-09-08 10:59:25.568022-0400 Pointer[26218:8830249] VERBOSE: getNotificationTypes:mSubscriptionStatus: -1
2018-09-08 10:59:25.569691-0400 Pointer[26218:8830249] VERBOSE: getNotificationTypes:mSubscriptionStatus: -1
2018-09-08 10:59:25.572585-0400 Pointer[26218:8830377] VERBOSE: getNotificationTypes:mSubscriptionStatus: -1
```

**Environment**

**Steps to Reproduce Issue:**

1. Install iOS SDK version 2.8.5 then update it to 2.8.7 with Cocoapods into a project
2. Initialize the SDK in didFinishLaunchingWithOptions:
3. Attempt to on getting playerId after granting push notification permissions

**Comment by Nightsd01** [comment]
@LamourBt Are you testing on simulator? If so, the simulator _can_ show the notification permission prompt but notifications don’t actually work in simulator (Apple restriction, not ours). 

If you are seeing this in a real device, there’s one thing to note. When a user grants permission, if it’s within 30 seconds of first launching the app, the SDK won’t immediately fire the subscription changed observer. First it has to get a OneSignal userID. So it first registers the user, _then_ it will fire the subscription observer. 

If the user hasn’t given/denied permission within 30 seconds of app first launch though, the SDK swill give up waiting and will register anyways (to get a user ID even if it can’t actually send notifications). 

Long story short, it may take a few seconds after the user gives permission to get the subscription observer to fire, or longer if they’re on a slow connection. 

Can you post a link to a full log file, from start to finish (app launch to at least 20 seconds after giving permission)?

**Comment by Nightsd01** [comment]
Also, can you double check your App ID matches the bundle ID for your app (ie. `com.myFlappyBirdClone.test`), and check to make sure the push notifications capability is enabled?

Several things could cause a symptom like this so let’s check the simple stuff first

**Comment by LamourBt** [comment]
@Nightsd01 Hey thanks for the response and no worry on the delay.  

I'm testing this on a real device not the simulator. 

The issue is that even after *granting permission access*, the log shows that *APNS* and *OneSignal Server* have the new playerId, it just the delegate now firing at all while the app is running. Only when closing and reopening the app the delegate will be fired and get the the info needed.

**Comment by Nightsd01** [comment]
@LamourBt That is very interesting. I am not able to reproduce this issue so I would be interested in trying to make my environment as similar to yours as possible.

If you have a few seconds I would appreciate it if you could answer some questions:

1. Where are you implementing your `OSSubscriptionObserver` delegate/protocol, and where in your code are you calling `OneSignal.addSubscriptionObserver`? Are you implementing it with your app delegate or somewhere else? If you are implementing it with an instance that is not the AppDelegate, could it possibly be that your observer object is getting released prematurely?

2. Can you reproduce this issue with one of our example projects (either ObjC or Swift, up to you)?

3. How are you triggering the permission prompt? Are you overriding `kOSSettingsKeyAutoPrompt`?

4. Are you able to successfully receive push notifications on the same device?

**Comment by Nightsd01** [comment]
Closing due to no response. If you're still encountering this issue please respond and I'll be happy to reopen and investigate.
